### PR TITLE
Add upper_bound kernel

### PR DESCRIPTION
**Pull request by aprokop**
- Add `KokkosExt::upper_bound` kernel and tests
- Move KokkosExt tests out of `ArborX_Test_DetailsUtils`

### CODE REVIEW

File: test/tstDetailsKokkosExtKernelStdAlgorithms.cpp [hunk 1]
```diff
@@ -41,13 +45,10 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(nth_element, DeviceType, ARBORX_DEVICE_TYPES)
   {
     int const n = v_ref.size();
 
-    Kokkos::View<float *, DeviceType> v("v", n);
-    Kokkos::deep_copy(
-        space, v,
-        Kokkos::View<float *, Kokkos::HostSpace,
-                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>(v_ref.data(), n));
+    Kokkos::View<float *, DeviceType> v("Testing::v", n);
+    Kokkos::deep_copy(space, v, UnmanagedView<float>(v_ref.data(), n));
```
Comment (dalg24) [inline test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:49]: Why do we care to copy with a specific exec space if we constructed the view without it?
Can't just use create_mirror_view_and_copy() ?

File: test/tstDetailsKokkosExtKernelStdAlgorithms.cpp [hunk 2]
```diff
@@ -65,3 +66,42 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(nth_element, DeviceType, ARBORX_DEVICE_TYPES)
...
+int upperBound(std::vector<float> const &v_host, float x)
+{
+  typename DeviceType::execution_space space;
+
+  auto const n = v_host.size();
+  Kokkos::View<float *, DeviceType> v("Testing::v", n);
+
+  Kokkos::deep_copy(space, v, UnmanagedView<float const>(v_host.data(), n));
+
+  int result;
```
Comment (dalg24) [inline test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:80]: Call it pos or position?

File: test/tstDetailsKokkosExtKernelStdAlgorithms.cpp [hunk 3]
```diff
@@ -65,3 +66,42 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(nth_element, DeviceType, ARBORX_DEVICE_TYPES)
     BOOST_TEST(nth_host == v_ref, tt::per_element());
   }
 }
+
+template <typename DeviceType>
+int upperBound(std::vector<float> const &v_host, float x)
```
Comment (dalg24) [inline test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:71]: Call it upperBoundPosition or findUpperBound (I prefer the latter)

File: test/CMakeLists.txt
[the remark this answers is not in the document]
Comment (aprokop) [reply test/CMakeLists.txt:1]: Updated the description.

**Comment by aprokop** [comment]
Addressed the comments, and rebased on `master` to resolve conflict.

File: test/tstDetailsKokkosExtKernelStdAlgorithms.cpp [hunk 4]
```diff
@@ -65,3 +67,39 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(nth_element, DeviceType, ARBORX_DEVICE_TYPES)
...
+
+template <typename DeviceType>
+int findUpperBound(std::vector<float> const &v_host, float x)
+{
+  auto const n = v_host.size();
+  Kokkos::View<float *, DeviceType> v("Testing::v", n);
+  Kokkos::deep_copy(v, UnmanagedHostView<float const>(v_host.data(), n));
+
+  int pos;
+  Kokkos::parallel_reduce(
```
Comment (dalg24) [inline test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:79]:
```suggestion
  Kokkos::parallel_reduce("Test::upper_bound",
```

File: test/tstDetailsKokkosExtKernelStdAlgorithms.cpp [hunk 5]
```diff
@@ -41,13 +45,11 @@ BOOST_AUTO_TEST_CASE_TEMPLATE(nth_element, DeviceType, ARBORX_DEVICE_TYPES)
   {
     int const n = v_ref.size();
 
-    Kokkos::View<float *, DeviceType> v("v", n);
-    Kokkos::deep_copy(
-        space, v,
-        Kokkos::View<float *, Kokkos::HostSpace,
-                     Kokkos::MemoryTraits<Kokkos::Unmanaged>>(v_ref.data(), n));
+    Kokkos::View<float *, DeviceType> v("Testing::v", 0);
```
Comment (dalg24) [inline test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:48]: I thought we were going for "[ArborX]{Test,Example}::foo"

**Review by dalg24 — APPROVED** [review]
Only have minor comments regarding the labelling.  I would prefer these to be fixed prior to squash-merging.

File: test/tstDetailsKokkosExtKernelStdAlgorithms.cpp (hunk 5 above)
Comment (aprokop) [reply test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:48]: We use `"Testing::foo` and `"Example::foo`. We use `ArborXTest` namespaces.

Comment (dalg24) [reply test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:48]: Please Test~~ing~~ (does not need to be resolved here

Comment (aprokop) [reply test/tstDetailsKokkosExtKernelStdAlgorithms.cpp:48]: > Please Test~ing~ (does not need to be resolved here

Are you asking to change all `Testing::` to `Test::` in a separate PR? Or to `ArborXTest::`?

**Pull request merged**